exchanges/kucoin/kucoin_websocket.go

```diff
@@ -1777,3 +1580,147 @@ func (ku *Kucoin) CalculateAssets(topic string, cp currency.Pair) ([]asset.Item,
+		case marginPositionChannel, marginLoanChannel:
+			s.Asset = asset.Margin
+		}
+	}
+	ku.Features.Subscriptions = slices.DeleteFunc(ku.Features.Subscriptions, func(s *subscription.Subscription) bool {
+		switch s.Channel {
+		case "/contractMarket/level2Depth50", // Replaced by subsctiption.Orderbook for asset.All
+			"/contractMarket/tickerV2", // Replaced by subscription.Ticker for asset.All
+			"/margin/fundingBook":      // Deprecated and removed
+			return true
+		case subscription.AllTradesChannel:
+			return s.Asset == asset.Empty
+		}
+		return false
+	})
+	ku.Config.Features.Subscriptions = ku.Features.Subscriptions
+}
+
+// channelName returns the correct channel name for the asset
+func channelName(s *subscription.Subscription, a asset.Item) string {
+	switch s.Channel {
+	case subscription.TickerChannel:
+		if a == asset.Futures {
+			return futuresTickerChannel
+		}
+		return marketTickerChannel
+	case subscription.OrderbookChannel:
+		if a == asset.Futures {
+			return futuresOrderbookDepth5Channel
+		}
+		return marketOrderbookDepth5Channel // This does not require a REST request to get the orderbook.
+	case subscription.CandlesChannel:
+		return marketCandlesChannel // No support in GCT yet for Futures candles
+	case subscription.AllTradesChannel:
+		return marketMatchChannel // No support in GCT yet for Futures all trades
+	}
+	return s.Channel
+}
+
+// removeSpotFromMargin removes spot pairs from margin pairs in the supplied AssetPairs map for subscriptions to non-margin endpoints
+func removeSpotFromMargin(s *subscription.Subscription, ap map[asset.Item]currency.Pairs, spotPairs currency.Pairs) string {
+	if strings.HasPrefix(s.Channel, "/margin") {
+		return ""
+	}
+	if p, ok := ap[asset.Margin]; ok {
+		ap[asset.Margin] = p.Remove(spotPairs...)
+	}
+	return ""
+}
+
+// isSymbolChannel returns if the channel expects receive a symbol
+func isSymbolChannel(s *subscription.Subscription) bool {
+	switch channelName(s, s.Asset) {
+	case privateSpotTradeOrders, accountBalanceChannel, marginPositionChannel, spotMarketAdvancedChannel, futuresSystemAnnouncementChannel,
+		futuresTradeOrderChannel, futuresStopOrdersLifecycleEventChannel, futuresAccountBalanceEventChannel:
+		return false
+	}
+	return true
+}
+
+// isCurrencyChannel returns if the channel expects receive a currency
+func isCurrencyChannel(s *subscription.Subscription) bool {
+	return s.Channel == marginLoanChannel
+}
+
+// channelInterval returns the channel interval if it has one
+func channelInterval(s *subscription.Subscription) string {
+	if channelName(s, s.Asset) == marketCandlesChannel {
+		if i, err := intervalToString(s.Interval); err == nil {
+			return i
+		}
+	}
+	return ""
+}
+
+// currencies returns the currencies from all pairs in an asset
+// Updates the AssetPairs map parameter to contain only those currencies as Base items
+func assetCurrencies(s *subscription.Subscription, ap map[asset.Item]currency.Pairs) currency.Currencies {
```

Comment:
Up to you, but I think it would be nice for the commentary to say something about how `ap` is generated in `fillAssetPairs` `/subscriptions/list.go`. It doesn't have a direct line from there to here

Also `currencies`->`assetCurrencies`

Reply:
I think that's okay and right, because of separation of responsibilities.
It's not down to anything below `expandTemplates` to know *how* it got the maps.
Especially since we're definitely going to move that.

But I did see that it's not clear *why* we're updating the AssetPairs map, so I've documented that.

Fixed 9b85f620

Reply:
Yeah that's fair 👍